<center><p float="center">
  <img src="https://upload.wikimedia.org/wikipedia/commons/e/e9/4_RGB_McCombs_School_Brand_Branded.png" width="300" height="100"/>
  <img src="https://mma.prnewswire.com/media/1458111/Great_Learning_Logo.jpg?p=facebook" width="200" height="100"/>
</p></center>

<h1><center><font size=10>Artificial Intelligence and Machine Learning</center></font></h1>



<center> <font size=6> Project Python Foundations: FoodHub Data Analysis

<center><img src="https://www.netsolutions.com/wp-content/uploads/2022/10/essential-features-of-building-an-on-demand-food-ordering-app.jpg" width="720"></center>


### Context

The number of restaurants in New York is increasing day by day. Lots of students and busy professionals rely on those restaurants due to their hectic lifestyles. Online food delivery service is a great option for them. It provides them with good food from their favorite restaurants. A food aggregator company FoodHub offers access to multiple restaurants through a single smartphone app.

The app allows the restaurants to receive a direct online order from a customer. The app assigns a delivery person from the company to pick up the order after it is confirmed by the restaurant. The delivery person then uses the map to reach the restaurant and waits for the food package. Once the food package is handed over to the delivery person, he/she confirms the pick-up in the app and travels to the customer's location to deliver the food. The delivery person confirms the drop-off in the app after delivering the food package to the customer. The customer can rate the order in the app. The food aggregator earns money by collecting a fixed margin of the delivery order from the restaurants.

### Objective

The food aggregator company has stored the data of the different orders made by the registered customers in their online portal. They want to analyze the data to get a fair idea about the demand of different restaurants which will help them in enhancing their customer experience. Suppose you are hired as a Data Scientist in this company and the Data Science team has shared some of the key questions that need to be answered. Perform the data analysis to find answers to these questions that will help the company to improve the business.

### Data Description

The data contains the different data related to a food order. The detailed data dictionary is given below.

### Data Dictionary

* order_id: Unique ID of the order
* customer_id: ID of the customer who ordered the food
* restaurant_name: Name of the restaurant
* cuisine_type: Cuisine ordered by the customer
* cost_of_the_order: Cost of the order
* day_of_the_week: Indicates whether the order is placed on a weekday or weekend (The weekday is from Monday to Friday and the weekend is Saturday and Sunday)
* rating: Rating given by the customer out of 5
* food_preparation_time: Time (in minutes) taken by the restaurant to prepare the food. This is calculated by taking the difference between the timestamps of the restaurant's order confirmation and the delivery person's pick-up confirmation.
* delivery_time: Time (in minutes) taken by the delivery person to deliver the food package. This is calculated by taking the difference between the timestamps of the delivery person's pick-up confirmation and drop-off information

### Let us start by importing the required libraries

In [3]:
# Installing the libraries with the specified version.
!pip install numpy==2.0.2 pandas==2.2.2 matplotlib==3.10.0 seaborn==0.13.2 -q --user

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [4]:
# import libraries for data manipulation
import numpy as np
import pandas as pd

# import libraries for data visualization
import matplotlib.pyplot as plt
import seaborn as sns

### Understanding the structure of the data

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [21]:
# Reading FoodHub order transaction history data set from google drive.
# Mounting and loading the order data in Panda DataFrame for further analytics

df = pd.read_csv("/content/drive/MyDrive/AIML-UTA-PY/foodhub_order.csv")

In [20]:
# Displaying the first 20 rows of the DataFrame
# This helps in validating that the data has been loaded correctly
# and provides an initial view of the data structure, format, schema and quality
# to help in the data analysis.
# Default is 5 rows, however, here I am displaying 20 rows

df.head(20)

,order_id,customer_id,restaurant_name,cuisine_type,cost_of_the_order,day_of_the_week,rating,food_preparation_time,delivery_time
0,1477147,337525,Hangawi,Korean,30.75,Weekend,Not given,25,20
1,1477685,358141,Blue Ribbon Sushi Izakaya,Japanese,12.08,Weekend,Not given,25,23
2,1477070,66393,Cafe Habana,Mexican,12.23,Weekday,5,23,28
3,1477334,106968,Blue Ribbon Fried Chicken,American,29.20,Weekend,3,25,15
4,1478249,76942,Dirty Bird to Go,American,11.59,Weekday,4,25,24
5,1477224,147468,Tamarind TriBeCa,Indian,25.22,Weekday,3,20,24
6,1477894,157711,The Meatball Shop,Italian,6.07,Weekend,Not given,28,21
7,1477859,89574,Barbounia,Mediterranean,5.97,Weekday,3,33,30
8,1477174,121706,Anjappar Chettinad,Indian,16.44,Weekday,5,21,26
9,1477311,39705,Bukhara Grill,Indian,7.18,Weekday,5,29,26


#### **Observations**
1. The dataset structure aligns well with the provided data dictionary, with all 20 columns present as expected.
2. Each row represents a unique food order, and order_id appears to act as a primary/unique identifier.
3. The dataset includes a mix of categorical variables (e.g., restaurant_name, cuisine_type, day_of_the_week) and numerical/decimal variables (e.g., cost_of_the_order, food_preparation_time, delivery_time).
4. The rating column contains both numerical values and text entries (e.g., “Not given”), indicating that special handling will be required before performing any numerical or mathematical analysis on this column.
5. Some restaurant names include Unicode characters (e.g., Chinese/Mandarin, observed in order_id 1477895), which indicates multilingual support and may require attention during text-based operations.
6. The data is indexed sequentially (0 to n), which is the default structure of a Pandas DataFrame.
7. Assuming the cost unit or currency is dollar ($) and refer it accordingly goring forward.


### **Question 1:** How many rows and columns are present in the data? [0.5 mark]

In [22]:
# Inspect the share of the dataset (rows, columns)
df.shape

(1898, 9)

#### Observations:
The dataset contains **1898** rows and **9** columns. This also mean total number of orders is 1898 in this dataset.

### **Question 2:** What are the datatypes of the different columns in the dataset? (The info() function can be used) [0.5 mark]

In [24]:
# Using DataFram info funcation inspecting data types, non-null counts, and
# overall structure of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1898 entries, 0 to 1897
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               1898 non-null   int64  
 1   customer_id            1898 non-null   int64  
 2   restaurant_name        1898 non-null   object 
 3   cuisine_type           1898 non-null   object 
 4   cost_of_the_order      1898 non-null   float64
 5   day_of_the_week        1898 non-null   object 
 6   rating                 1898 non-null   object 
 7   food_preparation_time  1898 non-null   int64  
 8   delivery_time          1898 non-null   int64  
dtypes: float64(1), int64(4), object(4)
memory usage: 133.6+ KB


#### Observations:
1. The dataset does not contain any null or missing values across the columns, indicating that all fields are populated. The Non-Null count matching with number or rows 1898.
2. Columns are integer identifiers order_id, customer_id, food_preparation_time, delivery_time are of integer data type
3. The Column cost_of_the_order is a float or decimal type which contains the cost.
4. Columns restaurant_name, cuisine_type, day_of_the_week are object data type.
5. The rating column is stored as an object data type due to the presence of both numerical values and text entries (e.g., "Not given"), and will require preprocessing before numerical analysis.

### **Question 3:** Are there any missing values in the data? If yes, treat them using an appropriate method. [1 mark]

In [29]:
# Checking for missing values in the dataset
df.isna().sum()


,0
order_id,0
customer_id,0
restaurant_name,0
cuisine_type,0
cost_of_the_order,0
day_of_the_week,0
rating,0
food_preparation_time,0
delivery_time,0


#### Observations:
1. There are missing data (null valuee) in the dataset
2. However, the rating column contains the value "Not given" (observed in data validation), is actually representing missing data.

### **Question 4:** Check the statistical summary of the data. What is the minimum, average, and maximum time it takes for food to be prepared once an order is placed? [2 marks]

In [30]:
# Generating descriptive statistics for numerical columns.
# df.describe() provides:
# - Count of non-missing values
# - Mean (average)
# - Standard deviation (spread of the data)
# - Min and max values
# - Quartiles (25%, 50%, 75%)
# This analysis matter more for 'cost_of_the_order',	'food_preparation_time'
# and	'delivery_time' column as ID columns are unique identifiers.
df.describe()

,order_id,customer_id,cost_of_the_order,food_preparation_time,delivery_time
count,1.898000e+03,1898.000000,1898.000000,1898.000000,1898.000000
mean,1.477496e+06,171168.478398,16.498851,27.371970,24.161749
std,5.480497e+02,113698.139743,7.483812,4.632481,4.972637
min,1.476547e+06,1311.000000,4.470000,20.000000,15.000000
25%,1.477021e+06,77787.750000,12.080000,23.000000,20.000000
50%,1.477496e+06,128600.000000,14.140000,27.000000,25.000000
75%,1.477970e+06,270525.000000,22.297500,31.000000,28.000000
max,1.478444e+06,405334.000000,35.410000,35.000000,33.000000


#### Observations:
1. The 'order_id' and 'customer_id' columns act as identifiers for each order and customer, respectively, and do not provide meaningful insights for numerical analysis.
2. The cost of an order ranges from 4.47 to 35.41 dollars (assuming dollar is the curency type), with an average order cost around 16 dollars and a standard deviation of 7.5 dollars.
3. The cost of 75% of the orders are below 23 dollars. This indicates that most of the customers prefer low-cost food compared to the expensive ones.
4. Food preparation time ranges from 20 to 35 minutes, with an average of around 27 minutes and a standard deviation of 4.6 minutes. The spread is not very high for the food preparation time.
5. Delivery time ranges from 15 to 33 minutes, with an average of around 24 minutes and a standard deviation of 5 minutes. The spread is not too high for delivery time either.

In [34]:
# Generate descriptive statistics for ALL columns (numeric + non-numeric)
# include='all' forces pandas to summarize:
# - Numeric columns (mean, std, min, quartiles, max)
# - Categorical columns (unique count, top value, frequency)
# - Boolean columns
# - Datetime columns (count, unique, min, max)
#
# .T transposes the output:
# - Rows become columns
# - Columns become rows

df.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
order_id,1898.0,NaN,NaN,NaN,1477495.5,548.049724,1476547.0,1477021.25,1477495.5,1477969.75,1478444.0
customer_id,1898.0,NaN,NaN,NaN,171168.478398,113698.139743,1311.0,77787.75,128600.0,270525.0,405334.0
restaurant_name,1898,178,Shake Shack,219,NaN,NaN,NaN,NaN,NaN,NaN,NaN
cuisine_type,1898,14,American,584,NaN,NaN,NaN,NaN,NaN,NaN,NaN
cost_of_the_order,1898.0,NaN,NaN,NaN,16.498851,7.483812,4.47,12.08,14.14,22.2975,35.41
day_of_the_week,1898,2,Weekend,1351,NaN,NaN,NaN,NaN,NaN,NaN,NaN
rating,1898,4,Not given,736,NaN,NaN,NaN,NaN,NaN,NaN,NaN
food_preparation_time,1898.0,NaN,NaN,NaN,27.37197,4.632481,20.0,23.0,27.0,31.0,35.0
delivery_time,1898.0,NaN,NaN,NaN,24.161749,4.972637,15.0,20.0,25.0,28.0,33.0


#### Observations:
1. The average order cost is approx 16.50 dollar, while the average food preparation time is around 27 minutes and the average delivery time is approximately 24 minutes.
2. 75th percentile shows that 75% of orders cost less than $22.30, indicating that most customers prefer moderately or low priced meals.
3. Food preparation time ranges from 20 minutes to 35 minutes, with an average of approximately 27 minutes, also with 75% of orders being prepared within 31 minutes, suggesting relatively consistent kitchen performance.
4. Delivery time ranges from 15 minutes to 33 minutes, with 75% of deliveries completed within 28 minutes, indicating stable delivery operations.
5. The standard deviation for cost (aprox 7.5) is relatively higher compared to preparation and delivery times (~4.6 and ~5 respectively), indicating greater variability in customer spending than in operational timings.
6. The variation in preparation and delivery times suggests that while most orders follow a consistent timeline, some orders take longer, which may impact customer experience.


In [35]:
# Calculate key statistics for food preparation time
# ---------------------------------------------------
# min()  -> shortest preparation time in the dataset
# mean() -> average preparation time across all orders
# max()  -> longest preparation time recorded
#
# These metrics help you understand:
# - Operational efficiency
# - Variability in kitchen performance
# - Potential bottlenecks in food preparation
food_prep_min = df['food_preparation_time'].min()
food_prep_mean = df['food_preparation_time'].mean()
food_prep_max = df['food_preparation_time'].max()
food_prep_min, food_prep_mean, food_prep_max

(20, np.float64(27.371970495258168), 35)

#### Observations:
1. The food preparation time ranges from 20 to 35 minutes, with an average of approximately 27 minutes, indicating generally consistent kitchen performance with moderate variation across orders.

### **Question 5:** How many orders are not rated? [1 mark]

In [60]:
# Get the total number of orders in the dataset
order_length = len(df)
print('Total Order Count:', order_length)
print()
# Filter the dataframe to include only rows where the rating is "Not given"
# Then count how many such orders exist
print("Rating     Count     %")
print("---------- --------- ----")
orders_without_rating = df[df['rating'] == 'Not given']['rating'].count()
print('Not Given  ' + str(orders_without_rating) + '       ' + f"{(orders_without_rating * 100 / order_length):.2f}" )
orders_with_rating_1 = df[df['rating'] == '1']['rating'].count()
print('1          ' + str(orders_with_rating_1) + '         ' + f"{(orders_with_rating_1 * 100 / order_length):.2f}" )
orders_with_rating_2 = df[df['rating'] == '2']['rating'].count()
print('2          ' + str(orders_with_rating_2) + '         ' + f"{(orders_with_rating_2 * 100 / order_length):.2f}" )
orders_with_rating_3 = df[df['rating'] == '3']['rating'].count()
print('3          ' + str(orders_with_rating_3) + '       ' + f"{(orders_with_rating_3 * 100 / order_length):.2f}" )
orders_with_rating_4 = df[df['rating'] == '4']['rating'].count()
print('4          ' + str(orders_with_rating_4) + '       ' + f"{(orders_with_rating_4 * 100 / order_length):.2f}" )
orders_with_rating_5 = df[df['rating'] == '5']['rating'].count()
print('5          ' + str(orders_with_rating_5) + '       ' + f"{(orders_with_rating_5 * 100 / order_length):.2f}" )


Total Order Count: 1898

Rating     Count     %
---------- --------- ----
Not Given  736       38.78
1          0         0.00
2          0         0.00
3          188       9.91
4          386       20.34
5          588       30.98


#### Observations:
1. A significant portion of orders (38.78%) have no rating,
2. Among the rated orders, most customers have given high ratings (4 and 5), indicating generally positive customer satisfaction.
3. Over 50% or ratings are 4 and 5
4. No customer gave rating 1 or 2

### Exploratory Data Analysis (EDA)

### Univariate Analysis

### **Question 6:** Explore all the variables and provide observations on their distributions. (Generally, histograms, boxplots, countplots, etc. are used for univariate exploration.) [9 marks]

In [ ]:
# Write the code here

### **Question 7**: Which are the top 5 restaurants in terms of the number of orders received? [1 mark]

In [ ]:
# Write the code here

#### Observations:


### **Question 8**: Which is the most popular cuisine on weekends? [1 mark]

In [ ]:
# Write the code here

#### Observations:


### **Question 9**: What percentage of the orders cost more than 20 dollars? [2 marks]

In [ ]:
# Write the code here

#### Observations:


### **Question 10**: What is the mean order delivery time? [1 mark]

In [ ]:
# Write the code here

#### Observations:


### **Question 11:** The company has decided to give 20% discount vouchers to the top 3 most frequent customers. Find the IDs of these customers and the number of orders they placed. [1 mark]

In [ ]:
# Write the code here

#### Observations:


### Multivariate Analysis

### **Question 12**: Perform a multivariate analysis to explore relationships between the important variables in the dataset. (It is a good idea to explore relations between numerical variables as well as relations between numerical and categorical variables) [10 marks]


In [ ]:
# Write the code here

### **Question 13:** The company wants to provide a promotional offer in the advertisement of the restaurants. The condition to get the offer is that the restaurants must have a rating count of more than 50 and the average rating should be greater than 4. Find the restaurants fulfilling the criteria to get the promotional offer. [3 marks]

In [ ]:
# Write the code here

#### Observations:


### **Question 14:** The company charges the restaurant 25% on the orders having cost greater than 20 dollars and 15% on the orders having cost greater than 5 dollars. Find the net revenue generated by the company across all orders. [3 marks]

In [ ]:
# Write the code here

#### Observations:


### **Question 15:** The company wants to analyze the total time required to deliver the food. What percentage of orders take more than 60 minutes to get delivered from the time the order is placed? (The food has to be prepared and then delivered.) [2 marks]

In [ ]:
# Write the code here

#### Observations:


### **Question 16:** The company wants to analyze the delivery time of the orders on weekdays and weekends. How does the mean delivery time vary during weekdays and weekends? [2 marks]

In [ ]:
# Write the code here

#### Observations:


### Conclusion and Recommendations

### **Question 17:** What are your conclusions from the analysis? What recommendations would you like to share to help improve the business? (You can use cuisine type and feedback ratings to drive your business recommendations.) [6 marks]

### Conclusions:
*  

### Recommendations:

*  

---